elfi/clients/multiprocessing.py

```diff
@@ -0,0 +1,106 @@
+import logging
+import itertools
+import multiprocessing
+
+from elfi.executor import Executor
+import elfi.client
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: use import hook instead? https://docs.python.org/3/reference/import.html
+def set_as_default():
+    elfi.client.set_client()
+    elfi.client.set_default_class(Client)
+
+
+class Client(elfi.client.ClientBase):
+    """
+    Client based on Python's built-in multiprocessing module.
+    
+    Parameters
+    ----------
+    num_processes : int, optional
+        Number of worker processes to use. Defaults to os.cpu_count().
+    """
+
+    def __init__(self, num_processes=None):
+        self.pool = multiprocessing.Pool(processes=num_processes)
+
+        self.tasks = {}
+        self._id_counter = itertools.count()
+
+    def apply(self, kallable, *args, **kwargs):
+        """Adds `kallable(*args, **kwargs)` to the queue of tasks. Returns immediately.
+        
+        Parameters
+        ----------
+        kallable : callable
+        
+        Returns
+        -------
+        id : int
+            Number of the queued task.
+        """
+        id = self._id_counter.__next__()
+        async_res = self.pool.apply_async(kallable, args, kwargs)
+        self.tasks[id] = async_res
+        return id
+
+    def apply_sync(self, kallable, *args, **kwargs):
+        """Calls and returns the result of `kallable(*args, **kwargs)`.
+        
+        Parameters
+        ----------
+        kallable : callable
+        """
+        return self.pool.apply(kallable, args, kwargs)
+
+    def get(self, task_id):
+        """Returns the result from task identified by `task_id` when it arrives.
+        
+        Parameters
+        ----------
+        task_id : int
+            Id of the task whose result to return.
+        """
+        async_result = self.tasks.pop(task_id)
+        return async_result.get()
+
+    def wait_next(self, task_ids):
+        for id in task_ids:
+            return self.get(id)
+
+    def is_ready(self, task_id):
+        """Return whether task with identifier `task_id` is ready.
+        
+        Parameters
+        ----------
+        task_id : int
+        """
+        return self.tasks[task_id].ready()
+
+    def remove_task(self, task_id):
+        """Remove task with identifier `task_id` from pool.
+        
+        Parameters
+        ----------
+        task_id : int
+        """
+        if task_id in self.tasks:
+            del self.tasks[task_id]
+        # TODO: also kill the pid?
+
+    def reset(self):
+        """Stop all worker processes immediately and clear pending tasks.
+        """
+        self.pool.terminate()
+        self.pool.join()
+        self.tasks.clear()
+
+    @property
+    def num_cores(self):
+        return self.pool._processes  # N.B. Not necessarily the number of actual cores.
```

Comment:
You mean the number of actual cores? That is the default. But a user may start the client with some other number of processes, so can't rely on cpu count here. :)